### ISSUE TITLE
Bug in immediate acquire logic

### ISSUE THREAD

**Issue by december1981** [Joshuaalbert/FairAsyncRLock#17]
Race condition:

when `event.set()` is called in the second snippet, another request to the lock can come in and immediately acquire it (first snippet), before the `await event.wait()` (third snippet) gets a chance to respond to the `event.set()`. So when this other `event.wait()` coro from the third snippet proceeds, it will clobber the owner and count for the immediately acquired coro from the first snippet. This leads to the exception  `f"Cannot release foreign lock. {me} tried to unlock {self._owner}."` when the first snippet coro duly releases.

```python
        # If the lock is free, acquire it immediately
        if self._count == 0:
            self._owner = me
            self._count = 1
            return
```

```python
    def _current_task_release(self):
        self._count -= 1
        if self._count == 0:
            self._owner = None
            if self._queue:
                # Wake up the next task in the queue
                event = self._queue.popleft()
                event.set()
```

```python
        try:
            await event.wait()
            self._owner = me
            self._count = 1
```

**Comment by Joshuaalbert** [comment]
Hi @december1981 thanks for reporting this. Let me make sure I understand.

Between the wake-up event of the next waiter in the queue being set, and the task which is waiting on that wake-up event getting its turn in the event loop, another task can execute first which claims the lock. 

This is a nice find, and the solution works.

The only issue is that I don't have a nice regression test so easily for this that fails before and passes after. Do you think you could make one that fails with pip version `1.0.6`?

In the mean time, I'll put this up as a PR.

**Comment by december1981** [comment]
> Between the wake-up event of the next waiter in the queue being set, and the task which is waiting on that wake-up event getting its turn in the event loop, another task can execute first which claims the lock.

Yep, that's right.

> The only issue is that I don't have a nice regression test so easily for this that fails before and passes after. Do you think you could make one that fails with pip version `1.0.6`?

Let me look into the unit tests then and I'll see if I can make one which reproduces this issue.

**Comment by december1981** [comment]
Tricky to get it to trigger, but got it eventually in task3()

```python
@pytest.mark.asyncio
async def test_fair_async_rlock_immediate_race_to_acquire():
    lock = FairAsyncRLock()

    # Use events to control the order of execution
    task1_aquired = asyncio.Event()
    # task1_released = asyncio.Event()
    task2_acquired = asyncio.Event()

    async def task1():
        async with lock:
            # tell task 2 to acquire lock
            task1_aquired.set()
            # but sleep long enough to make sure task 2 waits on lock before we release this one.
            await asyncio.sleep(0.1)
        # task1_released.set()

    async def task2():
        await task1_aquired.wait()
        async with lock:
            task2_acquired.set()
            # hold this for long enough to allow task3 to release first in a race condition
            # in which task3 beats task2 to ownership and then releases only to find
            # task2 has set ownership, giving a foreign lock assert scenario
            await asyncio.sleep(0.2)

    async def task3():
        # NB: we achieve this race condition with sleep(0.1)
        # awaiting task1_released does not give it the edge
        await asyncio.sleep(0.1)
        # await task1_released.wait()
        async with lock:
            # if task 3 were to get an immediate lock (beating task3 to ownership), then waiting
            # for this means we don't release until just after task2
            # has clobbered over our ownership.
            await task2_acquired.wait()

    t1 = asyncio.create_task(task1())
    t2 = asyncio.create_task(task2())
    t3 = asyncio.create_task(task3())

    await asyncio.gather(t1, t2, t3)
```

**Comment by december1981** [comment]
And to confirm, this new test asserts without my fix, and runs ok with it.

**Comment by december1981** [comment]
It dawned on me today that a much simpler fix is this:

```python
diff --git a/fair_async_rlock/fair_async_rlock.py b/fair_async_rlock/fair_async_rlock.py
index 51e6312..4d7be8b 100644
--- a/fair_async_rlock/fair_async_rlock.py
+++ b/fair_async_rlock/fair_async_rlock.py
@@ -42,7 +42,8 @@ class FairAsyncRLock:
         try:
             await event.wait()
             self._owner = me
-            self._count = 1
+            # count should be 1 from the waker event.set()
+            # assert self._count == 1
         except asyncio.CancelledError:
             try: # if in queue, then cancelled before release
                 self._queue.remove(event)
@@ -60,6 +61,7 @@ class FairAsyncRLock:
                 # Wake up the next task in the queue
                 event = self._queue.popleft()
                 event.set()
+                self._count = 1 # set to 1 now: prevent any immediate acquires before waiter resumes.
``` 

That also works with unit tests and my new test. I ran with the assert enabled to confirm also.

**Comment by december1981** [comment]
btw, without this in the test

```python
            # hold this for long enough to allow task3 to release first in a race condition
            # in which task3 beats task2 to ownership and then releases only to find
            # task2 has set ownership, giving a foreign lock assert scenario
            await asyncio.sleep(0.2)
```

If we don't sleep here, the other assert would fire without my fix: cannot release unacquired lock.

I guess sleeping is unnecessary there then, as the test exposes the issue either way.

**Comment by Joshuaalbert** [comment]
Okay, I'll update the PR

**Comment by Joshuaalbert** [comment]
Hmm, thinking about this. While it does make sense that setting _count=1 after the set() works (because we've ensured the queue is not empty before doing so), it makes reading the code more difficult. The previous approach reads better and has the same number of operations.

**Comment by december1981** [comment]
Up to you. Don't know if there's a meaningful performance cost to the extra `and not` in the immediate check, I would remark only.

**Comment by Joshuaalbert** [comment]
> Up to you. Don't know if there's a meaningful performance cost to the extra `and not` in the immediate check, I would remark only.

Thanks to python short-ciruiting the second part only gets evaluated in the case the first part `count == 0` is true (which would be rare -- only when the lock was free/this race condition)

**Comment by Joshuaalbert** [comment]
pip updated to 1.0.7 btw

**Comment by december1981** [comment]
count == 0 is a common case (cf when a fresh lock is attempted) and the and check will happen always along with it.
